### PR TITLE
Initial chunk data rewrite

### PR DESCRIPTION
**Pull request by NotStirred**
This PR attempts to rewrite the chunk system in a way that can be easily replaced for any other save formats.

I will readd the slightly faster (while loading) original version using this system in a later commit

### CODE REVIEW

File: chunky/src/java/se/llbit/chunky/world/Chunk.java [hunk 1]
```diff
@@ -220,15 +219,26 @@ private void loadBiomes(Map<String, Tag> data) {
...
+      }
     } else if (biomesTag.isIntArray(X_MAX * Z_MAX)) {
       // Since Minecraft 1.13, biome IDs are stored in an int vector.
       // TODO(llbit): do we need to use ints to store biome IDs for Minecraft 1.13+?
       int[] data = biomesTag.intArray();
-      for (int i = 0; i < X_MAX * Z_MAX; ++i) {
-        output[i] = (byte) data[i];
+      int i = 0;
+      for(int x = 0; x < X_MAX; x++) {
+        for(int z = 0; z < X_MAX; z++) {
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/Chunk.java]: Z_MAX

File: chunky/src/java/se/llbit/chunky/world/Chunk.java [hunk 2]
```diff
@@ -220,15 +219,26 @@ private void loadBiomes(Map<String, Tag> data) {
...
    * @param output    a byte array of length 16x16.
    */
-  private void extractBiomeData(@NotNull Tag biomesTag, byte[] output) {
+  private void extractBiomeData(@NotNull Tag biomesTag, ChunkData output) {
     if (biomesTag.isByteArray(X_MAX * Z_MAX)) {
-      System.arraycopy(biomesTag.byteArray(), 0, output, 0, X_MAX * Z_MAX);
+      byte[] data = biomesTag.byteArray();
+      int i = 0;
+      for(int x = 0; x < X_MAX; x++) {
+        for(int z = 0; z < X_MAX; z++) {
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/world/Chunk.java]: Z_MAX

File: chunky/src/java/se/llbit/chunky/chunk/ChunkData.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package se.llbit.chunky.chunk;
+
+import se.llbit.nbt.CompoundTag;
+
+import java.util.Collection;
+
+public interface ChunkData {
+  int blockAt(int x, int y, int z);
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/chunk/ChunkData.java]: `getBlockAt`

File: chunky/src/java/se/llbit/chunky/chunk/ChunkData.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+  void setBlockAt(int x, int y, int z, int block);
+
+  boolean blockOnEdge(int x, int y, int z);
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/chunk/ChunkData.java]: `isBlockOnEdge`

File: chunky/src/java/se/llbit/chunky/chunk/ChunkData.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  boolean isBlockOnEdge(int x, int y, int z);
+
+  Collection<CompoundTag> getTileEntities();
+
+  Collection<CompoundTag> getEntities();
+
+  byte getBiomeAt(int x, int y, int z);
+
+  void setBiomeAt(int x, int y, int z, byte biome);
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/chunk/ChunkData.java:24]: IIRC biomes are not bytes anymore

File: chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java [hunk 6]
```diff
@@ -825,13 +822,13 @@ public synchronized void loadChunks(TaskTracker progress, World world,
           int wz = cz + wz0;
           for (int cx = 0; cx < 16; ++cx) {
             int wx = cx + wx0;
-            int biomeId = 0xFF & biomes[Chunk.chunkXZIndex(cx, cz)];
+            int biomeId = 0xFF & chunkData.getBiomeAt(cx, 0, cz);
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java]: While you're at it, you might as well add a `// TODO add vertical biomes support (1.15+)`

File: chunky/src/java/se/llbit/chunky/chunk/ChunkData.java (hunk 5 above)
Comment (NotStirred) [reply chunky/src/java/se/llbit/chunky/chunk/ChunkData.java:24]: Chunky represented them as bytes previously, do you want me to move to `int[]` in this PR?

**Comment by NotStirred** [comment]
Sorry about the force pushes, I committed something I really shouldn't've

File: chunky/src/java/se/llbit/chunky/chunk/ChunkData.java (hunk 5 above)
Comment (leMaik) [reply chunky/src/java/se/llbit/chunky/chunk/ChunkData.java:24]: Looked into the Biomes class… byte[] is fine :+1:

**Comment by NotStirred** [comment]
closing as we agreed on discord to rebase onto master [here](https://discord.com/channels/541221265512464394/541221265512464397/796158339573612554)
New PR: https://github.com/chunky-dev/chunky/pull/773

**Pull request closed without merging**